rs/execution_environment/src/canister_manager.rs

```diff
@@ -2058,6 +2052,272 @@ impl CanisterManager {
+                let res = snapshot_inner
+                    .execution_snapshot_mut()
+                    .wasm_binary
+                    .write(&args.chunk, offset as usize);
+                if res.is_err() {
+                    return Err(CanisterManagerError::InvalidSubslice {
+                        offset,
+                        size: args.chunk.len() as u64,
+                    });
+                }
+                args.chunk.len() as u64
+            }
+            CanisterSnapshotDataOffset::MainMemory { offset } => {
+                let mut buffer = Buffer::new(snapshot_inner.wasm_memory().page_map.clone());
+                buffer.write(&args.chunk, offset as usize);
+                args.chunk.len() as u64
+            }
+            CanisterSnapshotDataOffset::StableMemory { offset } => {
+                let mut buffer = Buffer::new(snapshot_inner.stable_memory().page_map.clone());
+                buffer.write(&args.chunk, offset as usize);
+                args.chunk.len() as u64
+            }
+            CanisterSnapshotDataOffset::WasmChunk => {
+                // The chunk store is initialized as empty, and no memory cost has been charged yet.
+                // So we charge for both memory and instructions here.
+                let validated_chunk = match snapshot_inner
+                    .chunk_store_mut()
+                    .can_insert_chunk(self.config.wasm_chunk_store_max_size, args.chunk.clone())
+                {
+                    ChunkValidationResult::Insert(validated_chunk) => validated_chunk,
+                    ChunkValidationResult::AlreadyExists(hash) => {
+                        return Err(CanisterManagerError::WasmChunkStoreError {
+                            message: format!(
+                                "Chunk with hash {:02x?} already present in chunk store.",
+                                hash
+                            ),
+                        })
+                    }
+                    ChunkValidationResult::ValidationError(err) => {
+                        return Err(CanisterManagerError::WasmChunkStoreError { message: err })
+                    }
+                };
+
+                let memory_usage = canister.memory_usage();
+                let chunk_bytes = wasm_chunk_store::chunk_size();
+                let new_memory_usage = canister.memory_usage() + chunk_bytes;
+                let instructions = self.config.upload_wasm_chunk_instructions;
+                self.cycles_account_manager
+                    .consume_cycles_for_instructions(
+                        &sender,
+                        canister,
+                        instructions,
+                        subnet_size,
+                        // It does not matter if this is a Wasm64 or Wasm32 module.
+                        WasmExecutionMode::Wasm32,
+                    )
+                    .map_err(CanisterManagerError::CanisterSnapshotNotEnoughCycles)?;
```

Comment:
We must not fail after this: hence, this must be done as the last potentially failing operation (after memory_usage_checks).

Reply:
The reason is that this does check and update at the same time.